File: core/kill_switch.py
"""Hard kill-switch: halts all trading if daily drawdown exceeds threshold."""
from __future__ import annotations

import asyncio
import logging
from datetime import date, datetime

from prometheus_client import Counter, Gauge

from core.config import get_settings
from core.messaging import MsgType, get_bus
from core.models import PortfolioState

logger = logging.getLogger(__name__)

_kill_active = Gauge("kill_switch_active", "1 when kill switch is engaged")
_kill_triggers = Counter("kill_switch_triggers_total", "Total kill switch activations")


class KillSwitch:
    """
    Monitors daily drawdown. Trips automatically when loss exceeds
    max_daily_drawdown_pct. Resets at UTC midnight.
    """

    def __init__(self, initial_equity: float) -> None:
        cfg = get_settings()
        self.threshold = cfg.max_daily_drawdown_pct
        self._active = False
        self._peak_equity = initial_equity
        self._day_open_equity = initial_equity
        self._reset_date = date.today()

    @property
    def is_active(self) -> bool:
        return self._active

    def update(self, state: PortfolioState) -> bool:
        """Returns True if kill switch just triggered (newly activated)."""
        self._maybe_reset()
        if self._active:
            return False  # already tripped

        drawdown = (self._day_open_equity - state.equity) / self._day_open_equity
        if drawdown >= self.threshold:
            self._active = True
            _kill_active.set(1)
            _kill_triggers.inc()
            logger.critical(
                "KILL SWITCH ACTIVATED — daily drawdown %.2f%% >= threshold %.2f%%",
                drawdown * 100,
                self.threshold * 100,
            )
            return True
        return False

    def _maybe_reset(self) -> None:
        today = date.today()
        if today != self._reset_date:
            self._reset_date = today
            self._day_open_equity = (
                self._peak_equity  # conservative: use current equity as new baseline
            )
            logger.info("Kill switch day-reset — new baseline equity=%.2f", self._day_open_equity)

    def manual_reset(self) -> None:
        self._active = False
        _kill_active.set(0)
        logger.warning("Kill switch manually reset")

    async def broadcast_halt(self) -> None:
        bus = await get_bus()
        await bus.publish(
            "stream:system",
            MsgType.KILL_SWITCH,
            {"active": True, "timestamp": datetime.utcnow().isoformat()},
        )

File: sentinel-x/rust/src/risk/kill_switch.rs
//! Hard kill-switch: atomically trips on daily drawdown > threshold.
use prometheus::{Counter, Gauge, register_counter, register_gauge};
use std::sync::atomic::{AtomicBool, AtomicI64, Ordering};
use std::sync::Arc;
use std::time::{SystemTime, UNIX_EPOCH};

pub struct KillSwitch {
    active:           Arc<AtomicBool>,
    day_open_equity:  Arc<AtomicI64>,  // stored as fixed-point (×1_000_000)
    peak_equity:      Arc<AtomicI64>,
    threshold_pct:    f64,
    trigger_count:    Counter,
    active_gauge:     Gauge,
}

impl KillSwitch {
    pub fn new(initial_equity: f64, threshold_pct: f64) -> Self {
        let trigger_count = register_counter!(
            "kill_switch_triggers_total",
            "Total number of kill switch activations"
        )
        .unwrap();
        let active_gauge = register_gauge!(
            "kill_switch_active",
            "1 when the kill switch is engaged, 0 otherwise"
        )
        .unwrap();

        let fp = (initial_equity * 1_000_000.0) as i64;
        Self {
            active:          Arc::new(AtomicBool::new(false)),
            day_open_equity: Arc::new(AtomicI64::new(fp)),
            peak_equity:     Arc::new(AtomicI64::new(fp)),
            threshold_pct,
            trigger_count,
            active_gauge,
        }
    }

    /// Returns true if the kill switch just triggered (newly activated).
    /// Uses a compare-and-swap to prevent double-trips.
    pub fn check_and_trip(&self, current_equity: f64) -> bool {
        if self.active.load(Ordering::Acquire) {
            return false;  // already tripped
        }
        let day_open = self.day_open_equity.load(Ordering::Relaxed) as f64 / 1_000_000.0;
        let drawdown = (day_open - current_equity) / day_open;

        if drawdown >= self.threshold_pct {
            let was_inactive = self.active
                .compare_exchange(false, true, Ordering::AcqRel, Ordering::Relaxed)
                .is_ok();
            if was_inactive {
                self.trigger_count.inc();
                self.active_gauge.set(1.0);
                tracing::error!(
                    drawdown_pct = drawdown * 100.0,
                    threshold_pct = self.threshold_pct * 100.0,
                    "KILL SWITCH ACTIVATED — halting all trading"
                );
                return true;
            }
        }
        false
    }

    pub fn is_active(&self) -> bool {
        self.active.load(Ordering::Acquire)
    }

    pub fn manual_reset(&self, new_equity: f64) {
        let fp = (new_equity * 1_000_000.0) as i64;
        self.day_open_equity.store(fp, Ordering::Release);
        self.peak_equity.store(fp, Ordering::Release);
        self.active.store(false, Ordering::Release);
        self.active_gauge.set(0.0);
        tracing::warn!("Kill switch manually reset — new baseline equity={:.2}", new_equity);
    }

    pub fn current_drawdown_pct(&self, current_equity: f64) -> f64 {
        let day_open = self.day_open_equity.load(Ordering::Relaxed) as f64 / 1_000_000.0;
        ((day_open - current_equity) / day_open).max(0.0)
    }
}
